targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/README.md

```diff
@@ -0,0 +1,72 @@
+# Cortex-M3 Design Start Eval package example on MPS2+ board
+
+This folder includes the port of mbed OS on the example system of the Cortex-M3
+Design Start Eval package. This example is implemented in FPGA on the MPS2+
+board. Please check the [mbed page](https://os.mbed.com/platforms/ARM-CM3DS/) of
+this target for more information.
+For convenience, this target is called **CM3DS**.
+
+## Compiling :construction_worker:
+
+The target name is `ARM_CM3DS_MPS2`, you should be able to compile mbed OS
+projects for CM3DS with:
+
+```bash
+mbed compile -t COMPILER -m ARM_CM3DS_MPS2
+```
+
+The following compilers are supported (replace `COMPILER` with):
+
+* `ARM` for Arm Compiler version 5
+* `GCC_ARM` for GNU Compiler for Arm
+* `IAR` for IAR compiler
+
+## Running :running:
+
+Because of the new memory configuration introduced in commit `CM3DS: switch to
+larger memories for code and data` it
+has become more easy (and portable amoung all compilers) to use `.elf` files
+instead of `.bin`.
+For `.elf` files to work, you will need **at least version 2.2.5** of the MPS2+
+firmware. You can find firmware version 2.2.6 and instructions on how to put it
+in the MPS2+ board [here](https://community.arm.com/processors/designstart/f/discussions/9727/mps2-firmware-for-mbed).
+
+## Testing :zap:
+
+If you want to execute the mbed OS greentea tests on CM3DS, you will need
+**at least firmware version 2.2.6**.
+
+* `mbedls` does not automatically recognize which serial port is linked to the
+board. Check it manually and create a file named `mbedls.json` containing
+(at the same level than where you execute all commands):
+```bash
+{
+    "50040200074D652F3828F333": {
+        "serial_port": "/dev/ttyUSB0"
+    }
+}
+```
+Replace `/dev/ttyUSB0` with your correct serial port
+(something like `COM6` on Windows).
+
+* `mbedls` does not link CM3DS target ID with its name so execute the command:
+```bash
+mbedls --mock 5004:ARM_CM3DS_MPS2
+```
+
+* Because greentea works with `.bin` files, you will have to first compile the
+tests:
+```bash
+mbed test --compile -m ARM_CM3DS_MPS2 -t COMPILER
+```
+
+* Then use the script to change from `.bin` files to `.elf` in the test
+specification file:
+```bash
+./targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/scripts/bin_to_elf.py BUILD/tests/ARM_CM3DS_MPS2/COMPILER/test_spec.json
+```
+
+* You can now run the tests :+1: :
+```bash
+mbed test --run -m ARM_CM3DS_MPS2 -t COMPILER
+```
```

Comment:
@theotherjimmy I think that's a bit dodgy. Is this use (bin->elf) case already supported by our tools?

Reply:
Good catch.

@hug-dev From reading teh script, you just rename it. it's not real elf file is it? (I looked at function `bin_to_elf`).

> +Because of the new memory configuration introduced in commit `CM3DS: switch to
 +larger memories for code and data` it
 +has become more easy (and portable amoung all compilers) to use `.elf` files
 +instead of `.bin`.

What portability issues are you having here? Can you elaborate?

I would expect a target to support bin/hex files.

Reply:
@0xc0170 
It is real ELF files, the script changes the paths to use them instead on `.bin` files.
**More details are written in the commit message of the commit in that PR.**

The problem comes from that we are now loading at two different zones:

* the vector table at `0x0`, to allow the device to boot (boot address can not be changed)
* the rest of the code + data at address `0x00400000`

The ELF file produced contains the two load regions with the address information to where to load them.
The `.bin` file, however, is just plain code and data with no information at all which means that the single `.bin` contains everything with padding in between the end of the first load region and the beginnning of the second one. Consequence is that this `.bin` file will be huge, at least 4 MiB and hence the flash operation will take a much longer time than with ELF files.

@bulislaw 
I made sure that using ELF files were working with the testing tools, issuing the necessary pull requests or firmware changes of our target. The pull request ARMmbed/htrun#181, now merged works with the `bin_to_elf` script to use ELF files.

Reply:
The target does not support hex ? that would work out of the box.

Reply:
As far as I know, only `.bin` and `.elf` files are supported currently by the MPS2+ board firmware.

Reply:
@bulislaw No `.bin` to `.elf` is not a transformation we support. You could force the build system to produce an `.elf` file by setting `target.OUTPUT_EXT` to `"elf"` to skip the `.elf` -> `.bin` step, and therefore eliminate the `.bin` -> `.elf` step.

Reply:
Just to be clear, these patches do not add any transformation from `.bin` to `.elf`, they just make the test tools point to the `.elf` files (already produced with the compilation) instead of the `.bin` files.

Reply:
Yeah, I just read the code. You're renaming.